src/rust/src/rbackground.rs

```diff
@@ -0,0 +1,476 @@
+                        .as_function()
+                        .ok_or(RPolarsErr::new())
+                        .bad_val(rdbg(func_robj))
+                        .mistyped("pure R function")?;
+                    let shared_memory = serialize_series(RSeries::any_robj_to_pl_series_result(
+                        func.call(pairlist!(RSeries(series)))?,
+                    )?)?;
+                    RResult::Ok(ipc::IpcSharedMemory::from_bytes(shared_memory.as_slice()))
+                };
+                collector.send(bits().when(
+                    "trying to map a polars series with R function in the background R process",
+                ))?;
+                Ok(())
+            }
+        }
+    }
+}
+
+pub struct RBackgroundHandler {
+    proc: std::process::Child,
+    chan: ipc::IpcSender<RIPCJob>,
+}
+
+impl RBackgroundHandler {
+    pub fn new() -> RResult<Self> {
+        use std::process::*;
+
+        let (server, server_name) = ipc::IpcOneShotServer::new()
+            .when("trying to create a one-shot channel to setup inter-process communication")?;
+        let libs = RENV
+            .1
+            .iter()
+            .map(|lib| format!("\"{lib}\", "))
+            .collect::<Vec<_>>()
+            .join("");
+        let child = Command::new(RENV.0.as_str())
+            .arg("--vanilla")
+            .arg("-q")
+            .arg("-e")
+            // Remove rextendr::document() if possible
+            .arg(format!(
+                ".libPaths(c({libs}.libPaths())); polars:::handle_background_request(\"{server_name}\") |> invisible()"
+            ))
+            .stdin(Stdio::null())
+            .stdout(Stdio::null())
+            .stderr(Stdio::null())
+            .spawn()
+            .when("trying to spawn a background R process")?;
+        let (_, tx): (_, ipc::IpcSender<RIPCJob>) = server
+            .accept()
+            .when("waiting for the background R process to establish a job channel")?;
+        Ok(RBackgroundHandler {
+            proc: child,
+            chan: tx,
+        })
+    }
+
+    pub fn submit(&self, job: RIPCJob) -> RResult<()> {
+        self.chan
+            .send(job)
+            .when("trying to submit a job to the background R process")
+    }
+}
+
+impl Drop for RBackgroundHandler {
+    fn drop(&mut self) {
+        if self
+            .proc
+            .try_wait()
+            .ok()
+            .and_then(std::convert::identity)
+            .is_none()
+        {
+            self.proc.kill().unwrap();
+        }
+    }
+}
+
+pub struct RBackgroundPool {
+    pool: Arc<Mutex<VecDeque<RBackgroundHandler>>>,
+    cap: Arc<Mutex<usize>>,
+}
+
+impl RBackgroundPool {
+    pub fn new(cap: usize) -> Self {
+        RBackgroundPool {
+            pool: Arc::new(Mutex::new(VecDeque::new())),
+            cap: Arc::new(Mutex::new(cap)),
+        }
+    }
+
+    pub fn lease(&self) -> RResult<RBackgroundHandler> {
+        let mut sleep_time_us: u64 = 200;
```

Comment:
@Sicheng-Pan  new tentative lease method

Reply:
I have a smarter suggestion comming up :)

Reply:
Also for the current implementation I suppose it should always create new process when the pool is empty, just like before?